src/main/java/de/tum/in/www1/artemis/service/quiz/QuizScheduleService.java

```diff
@@ -0,0 +1,275 @@
+                if (quizBatch.get().getStartTime().isAfter(ZonedDateTime.now())) {
+                    scheduleService.scheduleTask(quizExercise, quizBatch.get(), ExerciseLifecycle.START, () -> executeQuizStartNowTask(quizExercise.getId()));
+                }
+            }
+        }
+        scheduleCalculateAllResults(quizExercise);
+    }
+
+    /**
+     * schedule the calculation of the results for all participations of the given quiz exercise
+     *
+     * @param quizExercise the quiz exercise for which the results should be calculated
+     */
+    public void scheduleCalculateAllResults(QuizExercise quizExercise) {
+        if (quizExercise.getDueDate() != null && !quizExercise.isQuizEnded()) {
+            // we only schedule the task if the quiz is not over yet
+            scheduleService.scheduleTask(quizExercise, ExerciseLifecycle.DUE,
+                    Set.of(new Tuple<>(quizExercise.getDueDate().plusSeconds(5), () -> calculateAllResults(quizExercise.getId()))));
+        }
+    }
+
+    /**
+     * Calculate the results for all participations of the given quiz exercise
+     *
+     * @param quizExerciseId the id of the quiz exercise for which the results should be calculated
+     */
+    public void calculateAllResults(long quizExerciseId) {
+        QuizExercise quizExercise = quizExerciseRepository.findByIdWithQuestionsAndStatisticsElseThrow(quizExerciseId);
+        log.info("Calculating results for quiz {}", quizExercise.getId());
+        studentParticipationRepository.findByExerciseId(quizExercise.getId()).forEach(participation -> {
+            participation.setExercise(quizExercise);
+            Optional<QuizSubmission> quizSubmissionOptional = quizSubmissionRepository.findWithEagerSubmittedAnswersByParticipationId(participation.getId());
+
+            if (quizSubmissionOptional.isEmpty()) {
+                return;
+            }
+            QuizSubmission quizSubmission = quizSubmissionOptional.get();
+
+            if (quizSubmission.isSubmitted()) {
+                if (quizSubmission.getType() == null) {
+                    quizSubmission.setType(SubmissionType.MANUAL);
+                }
+            }
+            else if (quizExercise.isQuizEnded()) {
+                quizSubmission.setSubmitted(true);
+                quizSubmission.setType(SubmissionType.TIMEOUT);
+                quizSubmission.setSubmissionDate(ZonedDateTime.now());
+            }
+
+            Result result = new Result().participation(participation);
+            result.setRated(true);
+            result.setAssessmentType(AssessmentType.AUTOMATIC);
+            result.setCompletionDate(quizSubmission.getSubmissionDate());
+            result.setSubmission(quizSubmission);
+
+            quizSubmission.calculateAndUpdateScores(quizExercise.getQuizQuestions());
+            result.evaluateQuizSubmission();
+
+            quizSubmissionRepository.save(quizSubmission);
+            resultRepository.save(result);
```

Comment:
I guess this could be improved by using `resultRepository.saveAll(..)` after the loop, to reduce the amount of calls to the database

Reply:
In principle yes, but then all savings would fail in case there is one problematic submission/result. So there is a tradeoff here. We should definitely involve error handling for the save calls in a follow up